stdlib/asm/collections/smt.masm

```diff
@@ -0,0 +1,249 @@
+#! Produces a remaining path key and index for depth 16
+#!
+#! Input:  [v, ...]
+#! Output: [(v << 16) >> 16, v >> 48, ...]
+#!
+#! Cycles: 9
+proc.shr_split_16
```

Comment:
Should we call this just `split_16`? I'm not sure bit shifting is the main operation here - but rather we are splitting the value into high 16 bits and low 48 bits.

Same comment applies to `shr_split_48`.

Reply:
Also, I think this is actually 7 cycles rather than 9.

Reply:
The root of that imprecision is probably something I could improve on when I count the cycles. Currently, I'm doing that manually; meaning: I check the docs of the instruction, and sum them up

The routine is implemented as follows
```
u32split
u32unchecked_divmod.65536
mul.4294967296
movup.2
add
```

- `u32split` is documented with `1` cycle
- `u32unchecked_divmod` is documented with `2-3` cycles; I arbitrarily picked `3` there
- `mul.b` is documented with `2` cycles
- `movup.n` is documented with `1-4` cycles; I arbitrarily picked `2` there
- `add` is documented with `1` cycle

I'm quite sure there is a more efficient way of doing that. This is error proning as I'm basically "guessing" from the ranges in the documentation.

Regarding the comment; fixed

Reply:
Yeah, we should have a good way to look up cycle count for each concrete instruction - but for now, I usually just look at the code (which is not the best way). Here are the actual cycle counts for the instructions where your estimate was off:

* `u32unchecked_divmod` will take 2 cycles in this case (it only takes 3 cycles if the immediate parameter is `1`).
* `movup.2` takes 1 cycle (it takes more than 1 cycle only for $n > 8$).